lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,231 @@
+require 'set'
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :visited
+
+        def initialize(template, files = nil)
+          # Serialize the template hash to deep-copy it and
+          #  avoid modifying the argument. Instead create a
+          #  new one to be modified by get_file_contents.
+          @template = deep_copy(template)
```

Comment:
No need to deep_copy. Why do we expect template can be a Hash? Initially it is either the content string or url.

Reply:
According to https://github.com/fog/fog-openstack/blob/master/docs/orchestration.md#stacks
stack_update accepts Hash too.

```
# from the above link.
heat_template = { "template": { "description": "Updated description" } }
stack.save(heat_template)

```
This allows the code to work in both cases.

Reply:
Ok, we agreed on irc to leave the deepcopy and support Hash to support orchestration.md examples.